pkg/plugins/policies/core/rules/rules.go

```diff
@@ -233,73 +279,107 @@ func BuildSingleItemRules(matchedPolicies []core_model.Resource) (SingleItemRule
-		ss := iter.Next()
-		if ss == nil {
-			break
+	// 3. Construct rules for all connected components of the graph independently
+	components := topo.ConnectedComponents(g)
+
+	for _, nodes := range components {
+		tagSet := map[Tag]bool{}
+		for _, node := range nodes {
+			for _, t := range subsets[node.ID()] {
+				tagSet[t] = true
+			}
 		}
-		// 3. For each combination determine a configuration
-		confs := []interface{}{}
-		// confs := []PolicyConf{}
-		distinctOrigins := map[core_model.ResourceKey]core_model.ResourceMeta{}
-		for i := 0; i < len(list); i++ {
-			item := list[i]
-			itemSubset, err := asSubset(item.GetTargetRef())
+
+		tags := []Tag{}
+		for tag := range tagSet {
+			tags = append(tags, tag)
+		}
+
+		sort.Slice(tags, func(i, j int) bool {
+			if tags[i].Key != tags[j].Key {
+				return tags[i].Key < tags[j].Key
+			}
+			return tags[i].Value < tags[j].Value
+		})
+
+		// 4. Iterate over all possible combinations with negations
+		iter := NewSubsetIter(tags)
+		for {
+			ss := iter.Next()
+			if ss == nil {
+				break
+			}
+			// 3. For each combination determine a configuration
+			confs := []interface{}{}
+			distinctOrigins := map[core_model.ResourceKey]core_model.ResourceMeta{}
+			for i := 0; i < len(list); i++ {
+				item := list[i]
+				itemSubset, err := asSubset(item.GetTargetRef())
+				if err != nil {
+					return nil, err
+				}
+				if itemSubset.IsSubset(ss) {
+					confs = append(confs, item.GetDefault())
+					distinctOrigins[core_model.MetaToResourceKey(item.ResourceMeta)] = item.ResourceMeta
+				}
+			}
+			merged, err := MergeConfs(confs)
 			if err != nil {
 				return nil, err
 			}
-			if itemSubset.IsSubset(ss) {
-				confs = append(confs, item.GetDefault())
-				distinctOrigins[core_model.MetaToResourceKey(item.ResourceMeta)] = item.ResourceMeta
+			if merged != nil {
+				var origins []core_model.ResourceMeta
+				for _, origin := range distinctOrigins {
+					origins = append(origins, origin)
+				}
+				sort.Slice(origins, func(i, j int) bool {
+					return origins[i].GetName() < origins[j].GetName()
+				})
+				rules = append(rules, &Rule{
+					Subset: ss,
+					Conf:   merged,
+					Origin: origins,
+				})
 			}
 		}
-		merged, err := MergeConfs(confs)
-		if err != nil {
-			return nil, err
-		}
-		if merged != nil {
-			var origins []core_model.ResourceMeta
-			for _, origin := range distinctOrigins {
-				origins = append(origins, origin)
-			}
-			sort.Slice(origins, func(i, j int) bool {
-				return origins[i].GetName() < origins[j].GetName()
-			})
-			rules = append(rules, &Rule{
-				Subset: ss,
-				Conf:   merged,
-				Origin: origins,
-			})
-		}
 	}
 
 	sort.SliceStable(rules, func(i, j int) bool {
-		return rules[i].Subset.NumPositive() > rules[j].Subset.NumPositive()
+		ss1, ss2 := rules[i].Subset, rules[j].Subset
+		if ss1.NumPositive() != ss2.NumPositive() {
+			return ss1.NumPositive() > ss2.NumPositive()
+		}
+		h1, _ := json.Marshal(ss1)
+		h2, _ := json.Marshal(ss2)
```

Comment:
I'm slightly concerned about perf there. Isn't it called multiple times for every DPP on every XDS reconciliation?

Reply:
Yeah surely there's a better way to do a comparison here :)